ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncActions.java

```diff
@@ -0,0 +1,131 @@
+package tech.pegasys.pantheon.ethereum.eth.sync.fastsync;
+
+import static tech.pegasys.pantheon.ethereum.eth.sync.fastsync.FastSyncError.CHAIN_TOO_SHORT;
+import static tech.pegasys.pantheon.ethereum.eth.sync.fastsync.FastSyncError.NO_PEERS_AVAILABLE;
+
+import tech.pegasys.pantheon.ethereum.ProtocolContext;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthScheduler;
+import tech.pegasys.pantheon.ethereum.eth.sync.SynchronizerConfiguration;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.PivotBlockRetriever;
+import tech.pegasys.pantheon.ethereum.eth.sync.tasks.WaitForPeerTask;
+import tech.pegasys.pantheon.ethereum.eth.sync.tasks.WaitForPeersTask;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+import tech.pegasys.pantheon.util.ExceptionUtils;
+
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FastSyncActions<C> {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private final SynchronizerConfiguration syncConfig;
+  private final ProtocolSchedule<C> protocolSchedule;
+  private final ProtocolContext<C> protocolContext;
+  private final EthContext ethContext;
+  private final LabelledMetric<OperationTimer> ethTasksTimer;
+
+  public FastSyncActions(
+      final SynchronizerConfiguration syncConfig,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+    this.syncConfig = syncConfig;
+    this.protocolSchedule = protocolSchedule;
+    this.protocolContext = protocolContext;
+    this.ethContext = ethContext;
+    this.ethTasksTimer = ethTasksTimer;
+  }
+
+  public CompletableFuture<Void> waitForSuitablePeers() {
+    final WaitForPeersTask waitForPeersTask =
+        WaitForPeersTask.create(
+            ethContext, syncConfig.getFastSyncMinimumPeerCount(), ethTasksTimer);
+
+    final EthScheduler scheduler = ethContext.getScheduler();
+    final CompletableFuture<Void> result = new CompletableFuture<>();
+    scheduler
+        .timeout(waitForPeersTask, syncConfig.getFastSyncMaximumPeerWaitTime())
+        .handle(
+            (waitResult, error) -> {
+              if (ExceptionUtils.rootCause(error) instanceof TimeoutException) {
+                if (ethContext.getEthPeers().bestPeer().isPresent()) {
+                  LOG.warn(
+                      "Fast sync timed out before minimum peer count was reached. Continuing with reduced peers.");
+                  result.complete(null);
+                } else {
+                  waitForAnyPeer()
```

Comment:
I think I would just have the action fail at this point and get handled at a higher level.

Reply:
I was thinking we'd just cut the `waitForAnyPeer()` part - seems a little iffy to continue with fast sync if we don't have enough peers.  Then the calling code could decide whether to run another round of waiting, or to abort fast sync.  Anyway - just thinking out loud.

Reply:
Yeah it's a bit of a toss up. The penalty for not waiting is a month long full sync so it's unlikely the user would want that, but maybe they'd want to be strict about waiting for a minimum number of peers.  

We probably should also allow the user to specifically set a pivot block by hash and we wait indefinitely until that block is available. That would be the most secure option since then users can guarantee they sync onto a chain they actually trust.